File: tests/integration/scripts/process_test.js
#!/usr/bin/env node
/**
 * Process operations test script for agent-jail integration testing.
 * Spawns subprocesses to verify process tree event capture.
 */
const { spawn, execSync, exec } = require('child_process');
const util = require('util');

const execAsync = util.promisify(exec);

async function main() {
    console.log("=== Node.js Process Test ===");
    console.log(`Main PID: ${process.pid}`);
    console.log(`Parent PID: ${process.ppid}`);

    // Spawn a simple subprocess (sync)
    console.log("\n--- execSync: echo ---");
    const output = execSync('echo "Hello from subprocess"', { encoding: 'utf-8' });
    console.log(`  stdout: ${output.trim()}`);

    // Async exec
    console.log("\n--- execAsync: shell command ---");
    const { stdout } = await execAsync('echo $SHELL && pwd');
    console.log(`  stdout: ${stdout.trim()}`);

    // Spawn with event handling
    console.log("\n--- spawn: ls ---");
    await new Promise((resolve) => {
        const ls = spawn('ls', ['-la', '/tmp']);
        let output = '';
        ls.stdout.on('data', data => output += data);
        ls.on('close', code => {
            console.log(`  Exit code: ${code}`);
            console.log(`  Output lines: ${output.trim().split('\n').length}`);
            resolve();
        });
    });

    // Multiple concurrent spawns
    console.log("\n--- Concurrent spawns ---");
    const procs = [];
    for (let i = 0; i < 3; i++) {
        const p = spawn('sleep', ['0.1']);
        console.log(`  Started sleep process ${i}: PID=${p.pid}`);
        procs.push(new Promise(resolve => p.on('close', code => {
            console.log(`  Process ${i} (PID=${p.pid}) exited with code ${code}`);
            resolve();
        })));
    }
    await Promise.all(procs);

    // Environment passing
    console.log("\n--- Environment passing ---");
    const result = await execAsync('echo TEST_VAR=$TEST_VAR', {
        env: { ...process.env, TEST_VAR: 'hello_world' }
    });
    console.log(`  ${result.stdout.trim()}`);

    // Working directory
    console.log("\n--- Working directory ---");
    const pwdResult = await execAsync('pwd', { cwd: '/tmp' });
    console.log(`  pwd in /tmp: ${pwdResult.stdout.trim()}`);

    // Pipe simulation
    console.log("\n--- Process pipe ---");
    const pipe1 = spawn('echo', ['-e', 'line1\\nline2\\nline3']);
    const pipe2 = spawn('wc', ['-l'], { stdio: [pipe1.stdout, 'pipe', 'pipe'] });
    let pipeOutput = '';
    pipe2.stdout.on('data', data => pipeOutput += data);
    await new Promise(resolve => pipe2.on('close', resolve));
    console.log(`  wc -l output: ${pipeOutput.trim()}`);

    console.log("\n=== Process Test Complete ===");
    return 0;
}

main().then(process.exit).catch(err => {
    console.error(err);
    process.exit(1);
});

File: tests/integration/scripts/process_test.py
#!/usr/bin/env python3
"""
Process operations test script for agent-jail integration testing.
Spawns subprocesses to verify process tree event capture.
"""
import os
import sys
import subprocess
import time

def main():
    print("=== Python Process Test ===")
    print(f"Main PID: {os.getpid()}")
    print(f"Parent PID: {os.getppid()}")

    # Spawn a simple subprocess
    print("\n--- Subprocess: echo ---")
    result = subprocess.run(
        ["echo", "Hello from subprocess"],
        capture_output=True,
        text=True
    )
    print(f"  stdout: {result.stdout.strip()}")
    print(f"  returncode: {result.returncode}")

    # Spawn subprocess with shell
    print("\n--- Subprocess: shell command ---")
    result = subprocess.run(
        "echo $SHELL && pwd",
        shell=True,
        capture_output=True,
        text=True
    )
    print(f"  stdout: {result.stdout.strip()}")

    # Spawn multiple concurrent subprocesses
    print("\n--- Concurrent subprocesses ---")
    procs = []
    for i in range(3):
        p = subprocess.Popen(
            ["sleep", "0.1"],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        )
        procs.append(p)
        print(f"  Started sleep process {i}: PID={p.pid}")

    # Wait for all to complete
    for i, p in enumerate(procs):
        p.wait()
        print(f"  Process {i} (PID={p.pid}) exited with code {p.returncode}")

    # Environment variable passing
    print("\n--- Environment passing ---")
    result = subprocess.run(
        ["sh", "-c", "echo TEST_VAR=$TEST_VAR"],
        env={**os.environ, "TEST_VAR": "hello_world"},
        capture_output=True,
        text=True
    )
    print(f"  {result.stdout.strip()}")

    # Working directory change
    print("\n--- Working directory ---")
    result = subprocess.run(
        ["pwd"],
        cwd="/tmp",
        capture_output=True,
        text=True
    )
    print(f"  pwd in /tmp: {result.stdout.strip()}")

    # Chain of processes (pipeline simulation)
    print("\n--- Process chain ---")
    p1 = subprocess.Popen(
        ["echo", "line1\nline2\nline3"],
        stdout=subprocess.PIPE
    )
    p2 = subprocess.Popen(
        ["wc", "-l"],
        stdin=p1.stdout,
        stdout=subprocess.PIPE
    )
    p1.stdout.close()
    output = p2.communicate()[0]
    print(f"  wc -l output: {output.decode().strip()}")

    print("\n=== Process Test Complete ===")
    return 0

if __name__ == "__main__":
    sys.exit(main())
